src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,309 @@
+
+size_t get_partition_size_from_heath_report(
+  const model::topic_namespace tp_ns,
+  const model::partition_id partition_id,
+  const cluster_health_report& health_report) {
+    for (auto node_report : health_report.node_reports) {
+        auto equal_topic = [&tp_ns](const topic_status& topic_status) {
+            return topic_status.tp_ns == tp_ns;
+        };
+        auto topic_status = std::find_if(
+          node_report.topics.begin(), node_report.topics.end(), equal_topic);
+        if (topic_status == node_report.topics.end()) {
+            continue;
+        }
+
+        auto equal_partition =
+          [&partition_id](const partition_status& partition_status) {
+              return partition_status.id == partition_id;
+          };
+        auto partition_status = std::find_if(
+          topic_status->partitions.begin(),
+          topic_status->partitions.end(),
+          equal_partition);
+        if (partition_status == topic_status->partitions.end()) {
+            continue;
+        }
+        return partition_status->size_bytes;
+        break;
+    }
+    vlog(
+      clusterlog.error,
+      "Partition {} {} status was not found in cluster health report",
+      tp_ns,
+      partition_id);
+    return 0;
+}
+
+result<allocation_units> partition_balancer_planner::get_reallocation(
+  const partition_assignment& assignments,
+  const topic_metadata& topic_metadata,
+  std::vector<model::broker_shard>& stable_replicas,
+  cluster_health_report& health_report,
+  size_t partition_size) {
+    auto stable_assigments = partition_assignment(
+      assignments.group, assignments.id, stable_replicas);
+
+    allocation_constraints allocation_constraints;
+
+    // TODO: add check if max_disc_usage enabled
+    // Add constraint on least disk usage
+    allocation_constraints.soft_constraints.push_back(
+      ss::make_lw_shared<soft_constraint_evaluator>(least_disk_filled(
+        health_report, _config.max_disc_usage, _assigned_reallocation_sizes)));
+
+    // Add constraint on partition disk overfill
+    allocation_constraints.hard_constraints.push_back(
+      ss::make_lw_shared<hard_constraint_evaluator>(
+        disk_not_overflowed_by_partition(
+          health_report,
+          _config.max_disc_usage,
+          partition_size,
+          _assigned_reallocation_sizes)));
+
+    // Add constraint on unavailable nodes
+    std::vector<model::broker_shard> unavailable_nodes;
+    for (auto node_id : _unavailable_nodes) {
+        unavailable_nodes.push_back(model::broker_shard{node_id, 0});
+    }
+    allocation_constraints.hard_constraints.push_back(
+      ss::make_lw_shared<hard_constraint_evaluator>(
+        distinct_from(unavailable_nodes)));
+
+    auto constraints = partition_constraints(
+      assignments.id,
+      topic_metadata.get_configuration().replication_factor,
+      allocation_constraints);
+
+    auto reallocation = _partition_allocator.local().reallocate_partition(
```

Comment:
we need to add some heuristic that checks if movement is unlikely to be successful (e.g. when a quorum of nodes is unavailable)

Reply:
fixed